janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java

```diff
@@ -2307,4 +2310,71 @@ public void testOrForceIndexComposite() throws Exception {
             }
         }
     }
+
+    /**
+     * This test builds a mixed index and tests index queries with order and limit.
+     * It also tests if index query cache is utilised correctly.
+     */
+    @Test
+    public void testOrderByWithLimit() {
+        final PropertyKey age = makeKey("age", Integer.class);
+        final JanusGraphIndex mixed = mgmt.buildIndex("mixed", Vertex.class).addKey(age).buildMixedIndex(INDEX);
+        finishSchema();
+
+        for (int i = 0; i < 100; i++) {
+            tx.addVertex("age", i);
+        }
+        tx.commit();
+
+        clopen(option(ADJUST_LIMIT),false);
+        final int[] ascTop100 = IntStream.range(0, 100).toArray();
+        final int[] ascTop30 = IntStream.range(0, 30).toArray();
+        final int[] ascTop10 = IntStream.range(0, 10).toArray();
+        final int[] descTop100 = IntStream.range(0, 100).map(i -> 99 - i).toArray();
+        final int[] descTop30 = IntStream.range(0, 30).map(i -> 99 - i).toArray();
+        final int[] descTop10 = IntStream.range(0, 10).map(i -> 99 - i).toArray();
+
+        Supplier<GraphTraversal<Vertex, Vertex>> traversal = () -> graph.traversal().V().has("age", P.gte(0)).order();
+
+        // traverse with limit 30 (cache cold miss)
+        assertTrue(hasBackendQuery((TraversalMetrics) traversal.get().by(ORDER_AGE_ASC).limit(30).values("age").profile().next()));
+        List<Integer> ascAges30 = traversal.get().by(ORDER_AGE_ASC).limit(30).values("age").toList();
+        assertArrayEquals(ascTop30, ascAges30.stream().mapToInt(i -> i).toArray());
+        assertTrue(hasBackendQuery((TraversalMetrics) traversal.get().by(ORDER_AGE_DESC).limit(30).values("age").profile().next()));
+        List<Integer> descAges30 = traversal.get().by(ORDER_AGE_DESC).limit(30).values("age").toList();
+        assertArrayEquals(descTop30, descAges30.stream().mapToInt(i -> i).toArray());
+
+        // traverse without limit (cache cold miss)
+        assertTrue(hasBackendQuery((TraversalMetrics) traversal.get().by(ORDER_AGE_ASC).values("age").profile().next()));
+        List<Integer> ascAges = traversal.get().by(ORDER_AGE_ASC).values("age").toList();
+        assertArrayEquals(ascTop100, ascAges.stream().mapToInt(i -> i).toArray());
+        assertTrue(hasBackendQuery((TraversalMetrics) traversal.get().by(ORDER_AGE_DESC).values("age").profile().next()));
```

Comment:
If no limit is given, the cached result can be used if ordering is applied afterwards. But this wasn't done before so I'm also ok with declaring it out of scope of this PR.

Reply:
The cached values are just vertex ids. Unless we also cache indexed field values (for mixed indexes, of course, since composite indexes do not support ordering query natively anyway), it's impossible to use caching results if ordering is applied afterward.